src/dune_engine/fs_memo.ml

```diff
@@ -2,10 +2,47 @@ open! Stdune
 open! Import
 open Memo.Build.O
 
+type t = { dune_file_watcher : Dune_file_watcher.t option }
+
+(* Ideally this should be an [Fdecl], but there are currently two reasons why
+   it's not:
+
+   - We read the workspace file before we start the inotify watcher, so [init]
+   is called after [t_ref] is used. This is clearly not good (it means we never
+   re-read the workspace file)
```

Comment:
Could we buffer requests until `init` is called?

Reply:
@jeremiedimino, that would probably be better than nothing, but then we'd still not be sensitive to the changes that happened between the time we read the file and the time we subscribed. I'll make that change since it's only makes things better, but I think to fix it completely we'll need one of two things:

- re-read all files after initialization
- only read the workspace file after initialization in the first place

Reply:
I pushed commit 9397fd0 in which I both do this and do the invalidation of the appropriate Memo nodes.
I think potentially fixes the "bug" completely, even though it's somewhat awkward.

Reply:
Reading the workspace file one more time seems acceptable as it's not that big. It should also allow to simplify workspace.ml since sharing the parsing of the `Dune_config.t` part is no longer useful.